File: packages/qf-scaffolding/src/utils/logger.ts
/**
 * Unified logging for qf-scaffolding (RFC-001 compliant)
 *
 * All logs go to logs/session.jsonl (shared with qf-pipeline).
 */

import * as fs from "fs";
import * as path from "path";
import * as yaml from "js-yaml";

// Schema version per RFC-001
const SCHEMA_VERSION = 1;

// Log levels
export type LogLevel = "debug" | "info" | "warn" | "error";

// Log entry structure per RFC-001
interface LogEntry {
  ts: string;
  v: number;
  session_id: string;
  mcp: string;
  tool: string;
  event: string;
  level: LogLevel;
  data?: Record<string, unknown>;
  duration_ms?: number;
  parent_id?: string;
}

// Session YAML structure
interface SessionYaml {
  session?: {
    id?: string;
    created?: string;
    updated?: string;
  };
}

/**
 * Read session_id from session.yaml
 */
function readSessionId(projectPath: string): string {
  try {
    const sessionYamlPath = path.join(projectPath, "session.yaml");
    if (fs.existsSync(sessionYamlPath)) {
      const content = fs.readFileSync(sessionYamlPath, "utf-8");
      const data = yaml.load(content, { schema: yaml.JSON_SCHEMA }) as SessionYaml;
      return data?.session?.id || "unknown";
    }
  } catch {
    // Ignore errors, return unknown
  }
  return "unknown";
}

/**
 * Log event to logs/session.jsonl (RFC-001 compliant)
 *
 * Thread-safe with synchronous writes.
 * If sessionId is empty, it will be auto-read from session.yaml.
 */
export function logEvent(
  projectPath: string,
  sessionId: string,
  tool: string,
  event: string,
  level: LogLevel = "info",
  data?: Record<string, unknown>,
  durationMs?: number,
  parentId?: string
): void {
  if (!projectPath) {
    return;
  }

  // Auto-read session_id if not provided
  const effectiveSessionId = sessionId || readSessionId(projectPath);

  // Ensure logs directory exists
  const logsDir = path.join(projectPath, "logs");
  if (!fs.existsSync(logsDir)) {
    fs.mkdirSync(logsDir, { recursive: true });
  }

  // Build log entry per RFC-001 schema
  const logEntry: LogEntry = {
    ts: new Date().toISOString(),
    v: SCHEMA_VERSION,
    session_id: effectiveSessionId,
    mcp: "qf-scaffolding",
    tool,
    event,
    level,
  };

  // Add optional fields
  if (data) {
    logEntry.data = data;
  }
  if (durationMs !== undefined) {
    logEntry.duration_ms = durationMs;
  }
  if (parentId) {
    logEntry.parent_id = parentId;
  }

  // Append to session.jsonl (shared log)
  const logFile = path.join(logsDir, "session.jsonl");
  fs.appendFileSync(logFile, JSON.stringify(logEntry) + "\n", "utf-8");
}

/**
 * Log action (backward-compatible wrapper)
 *
 * Automatically reads session_id from session.yaml.
 */
export function logAction(
  projectPath: string,
  tool: string,
  message: string,
  data?: Record<string, unknown>,
  level: LogLevel = "info"
): void {
  if (!projectPath) {
    return;
  }

  // Auto-read session_id from session.yaml
  const sessionId = readSessionId(projectPath);

  logEvent(projectPath, sessionId, tool, message, level, data);
}

/**
 * Log stage event (helper for M1-M4 stages)
 */
export function logStageEvent(
  projectPath: string,
  module: string,
  stage: number,
  event: "stage_start" | "stage_complete" | "stage_skip",
  data?: Record<string, unknown>
): void {
  const sessionId = readSessionId(projectPath);

  logEvent(
    projectPath,
    sessionId,
    `${module}_stage${stage}`,
    event,
    "info",
    {
      module,
      stage,
      ...data,
    }
  );
}

/**
 * Log user decision (for ML training)
 */
export function logUserDecision(
  projectPath: string,
  tool: string,
  decisionType: string,
  optionsPresented: unknown[],
  userChoice: unknown,
  rationale?: string
): void {
  const sessionId = readSessionId(projectPath);

  logEvent(
    projectPath,
    sessionId,
    tool,
    "user_decision",
    "info",
    {
      decision_type: decisionType,
      options_presented: optionsPresented,
      user_choice: userChoice,
      ...(rationale && { rationale }),
    }
  );
}

/**
 * Log error with context
 */
export function logError(
  projectPath: string,
  tool: string,
  errorType: string,
  errorMessage: string,
  context?: Record<string, unknown>
): void {
  const sessionId = readSessionId(projectPath);

  logEvent(
    projectPath,
    sessionId,
    tool,
    "tool_error",
    "error",
    {
      error_type: errorType,
      error_message: errorMessage,
      ...context,
    }
  );
}

/**
 * Get session state from logs (for resumption)
 */
export function getSessionState(projectPath: string): {
  status: string;
  total_events?: number;
  last_activity?: string;
  error_count?: number;
} {
  const logFile = path.join(projectPath, "logs", "session.jsonl");

  if (!fs.existsSync(logFile)) {
    return { status: "no_session" };
  }

  try {
    const content = fs.readFileSync(logFile, "utf-8");
    const lines = content.trim().split("\n").filter(Boolean);
    const events = lines.map((line) => JSON.parse(line) as LogEntry);

    if (events.length === 0) {
      return { status: "empty", total_events: 0 };
    }

    const errors = events.filter((e) => e.level === "error");

    return {
      status: "resumable",
      total_events: events.length,
      last_activity: events[events.length - 1]?.ts,
      error_count: errors.length,
    };
  } catch {
    return { status: "error" };
  }
}

File: packages/qf-pipeline/src/qf_pipeline/utils/logger.py
"""Pipeline logging utility.

Unified logging for QuestionForge (RFC-001).
All logs go to logs/session.jsonl (shared by qf-pipeline and qf-scaffolding).
"""

import fcntl
import json
from pathlib import Path
from datetime import datetime, timezone
from typing import Optional, Dict, Any

# Schema version per RFC-001
SCHEMA_VERSION = 1


def _read_session_id(project_path: Path) -> str:
    """Read session_id from session.yaml.

    Args:
        project_path: Project directory containing session.yaml

    Returns:
        Session ID string, or "unknown" if not found
    """
    try:
        import yaml
        session_yaml = project_path / "session.yaml"
        if session_yaml.exists():
            data = yaml.safe_load(session_yaml.read_text())
            return data.get("session", {}).get("id", "unknown")
    except Exception:
        pass
    return "unknown"


def log_action(
    project_path: Path,
    step: str,
    message: str,
    data: dict = None,
    level: str = "info"
):
    """Log action to logs/session.jsonl (backward-compatible wrapper).

    This is a wrapper around log_event for backwards compatibility.
    Automatically reads session_id from session.yaml.

    Args:
        project_path: Project directory containing log files
        step: Tool/step name (e.g., "step0_start", "step2_validate")
        message: Human-readable message (becomes 'event' field)
        data: Optional structured data for JSON log
        level: Log level (debug, info, warn, error). Default: "info"
    """
    if project_path is None:
        return

    project_path = Path(project_path)

    # Auto-read session_id from session.yaml
    session_id = _read_session_id(project_path)

    # Filter out 'project_path' from data to avoid duplication
    safe_data = {k: v for k, v in (data or {}).items() if k != 'project_path'}

    log_event(
        project_path=project_path,
        session_id=session_id,
        tool=step,
        event=message,
        level=level,
        data=safe_data if safe_data else None
    )


def log_event(
    project_path: Path,
    session_id: str,
    tool: str,
    event: str,
    level: str = "info",
    data: Optional[Dict[str, Any]] = None,
    duration_ms: Optional[int] = None,
    parent_id: Optional[str] = None,
    mcp: str = "qf-pipeline"
) -> None:
    """Log event to logs/session.jsonl (RFC-001 compliant).

    This is the shared logging format used by both qf-pipeline and qf-scaffolding.
    Thread-safe with file locking.

    Args:
        project_path: Project directory
        session_id: UUID linking all events in a session
        tool: Tool name (e.g., "step0_start", "load_stage")
        event: Event type (e.g., "session_start", "tool_end")
        level: Log level (debug, info, warn, error). Default: "info"
        data: Optional structured data
        duration_ms: Optional operation duration in milliseconds
        parent_id: Optional parent event ID for hierarchical logging
        mcp: MCP name (default: "qf-pipeline")
    """
    if project_path is None:
        return

    project_path = Path(project_path)
    logs_dir = project_path / "logs"
    logs_dir.mkdir(exist_ok=True)

    # Build log entry per RFC-001 schema
    log_entry = {
        "ts": datetime.now(timezone.utc).isoformat().replace('+00:00', 'Z'),
        "v": SCHEMA_VERSION,
        "session_id": session_id,
        "mcp": mcp,
        "tool": tool,
        "event": event,
        "level": level,
    }

    # Add optional fields
    if data:
        log_entry["data"] = data
    if duration_ms is not None:
        log_entry["duration_ms"] = duration_ms
    if parent_id:
        log_entry["parent_id"] = parent_id

    # Append to session.jsonl (shared log) with file locking
    session_log = logs_dir / "session.jsonl"
    with open(session_log, "a", encoding="utf-8") as f:
        fcntl.flock(f, fcntl.LOCK_EX)
        try:
            f.write(json.dumps(log_entry, ensure_ascii=False) + "\n")
        finally:
            fcntl.flock(f, fcntl.LOCK_UN)


def get_session_state(project_path: Path) -> Dict[str, Any]:
    """Read session.jsonl and determine current state for resumption.

    Args:
        project_path: Project directory

    Returns:
        Dict with session state info
    """
    log_file = Path(project_path) / "logs" / "session.jsonl"

    if not log_file.exists():
        return {"status": "no_session", "events": 0}

    events = []
    try:
        with open(log_file, encoding="utf-8") as f:
            for line in f:
                if line.strip():
                    events.append(json.loads(line))
    except Exception as e:
        return {"status": "error", "error": str(e)}

    if not events:
        return {"status": "empty", "events": 0}

    # Find last completed stage/tool
    last_complete = None
    for e in reversed(events):
        if e.get("event") in ("stage_complete", "tool_end"):
            if e.get("data", {}).get("success", True):
                last_complete = e
                break

    # Count errors
    errors = [e for e in events if e.get("level") == "error"]

    return {
        "status": "resumable",
        "total_events": len(events),
        "last_activity": events[-1].get("ts") if events else None,
        "last_complete": last_complete,
        "error_count": len(errors),
        "errors": errors[-5:] if errors else []  # Last 5 errors
    }
